mono/mini/alias-analysis.c

```diff
@@ -157,7 +157,44 @@ lower_memory_access (MonoCompile *cfg)
 	MonoInst *ins, *tmp;
 	gboolean needs_dce = FALSE;
 	GHashTable *addr_loads = g_hash_table_new (NULL, NULL);
-	//FIXME optimize
+	MonoInst **defs;
+
+	/*
+	 * Find all vregs whose only definition is an ldaddr opcode.
+	 */
+	defs = g_malloc0 (sizeof (MonoInst*) * cfg->next_vreg);
+	for (bb = cfg->bb_entry; bb; bb = bb->next_bb) {
+		for (ins = bb->code; ins; ins = ins->next) {
+			if (ins->dreg != -1) {
+				if (defs [ins->dreg])
+					/* Mark as having multiple defs */
+					defs [ins->dreg] = GINT_TO_POINTER (-1);
+				else
+					defs [ins->dreg] = ins;
+			}
+		}
+	}
+
+	for (int i = 0; i < cfg->next_vreg; ++i) {
+		if (defs [i]) {
+			if (defs [i] == GINT_TO_POINTER (-1) || defs [i]->opcode != OP_LDADDR)
+				defs [i] = NULL;
+		}
+	}
+
+	for (int i = 0; i < cfg->next_vreg; ++i) {
+		if (defs [i]) {
+			ins = defs [i];
+			MonoInst *var = (MonoInst*)ins->inst_p0;
+			if (var->flags & MONO_INST_VOLATILE) {
```

Comment:
This must check for MONO_INST_INDIRECT as well. But this would only makes sense if we do it after we do extract/insert conversion.

This needs to be tested with other sources of indirection such as passing/return from functions and taking a ref to a span.

Reply:
Extract is going to be translated back to a load in the vtype decomposition pass. So this is just a temporary instruction used to get rid of ldaddr opcodes, making the vtype non-indirect.

Reply:
The issue is that MONO_INST_INDIRECT has the actual effect you're trying to catch here. I did cleanup usage of MONO_INST_VOLATILE a few years ago to only mean volatile and not "address taken".

Reply:
This code tries to compute vregs whose only definition of OP_LDADDR, not anything more, so we can do ldaddr+load -> extract transformations.